test/unit_tests/arkode/C_serial/ark_reset_check.c

```diff
@@ -0,0 +1,451 @@
+  if (check_retval((void *)LS, "SUNLinSol_Dense", 0)) return 1;
+
+
+  /******* Part I: ERKStep *******/
+
+  /* Set initial condition, and construct stepper */
+  t = T0;
+  N_VConst(ytrue(t), y);
+  arkode_mem = ERKStepCreate(f, t, y, ctx);
+  if (check_retval((void *)arkode_mem, "ERKStepCreate", 0)) return 1;
+  retval = ERKStepSetUserData(arkode_mem, (void *) &lambda);
+  if (check_retval(&retval, "ERKStepSetUserData", 1)) return 1;
+  retval = ERKStepSStolerances(arkode_mem, rtol, atol);
+  if (check_retval(&retval, "ERKStepSStolerances", 1)) return 1;
+  retval = ERKStepSetMaxNumSteps(arkode_mem, 1000);
+  check_retval(&retval, "ERKStepSetMaxNumSteps", 1);
+
+  /* Initially evolve to dTout, and check result */
+  retval = ERKStepSetStopTime(arkode_mem, t + dTout);
+  check_retval(&retval, "ERKStepSetStopTime", 1);
+  retval = ERKStepEvolve(arkode_mem, t + dTout, y, &t, ARK_NORMAL);
+  if (check_retval(&retval, "ERKStepEvolve", 1)) return 1;
+  if (check_ans(y, t, RCONST(0.001), RCONST(0.000001))) {
+    printf("  Initial ERKStepEvolve had insufficient accuracy\n");
+    printf("    t = %" GSYM "\n", t);
+    printf("    y = %" GSYM "\n", NV_Ith_S(y,0));
+    printf("    ytrue = %" GSYM "\n", ytrue(t));
+    printf("    |y-ytrue| = %" GSYM "\n", SUNRabs(ytrue(t)-NV_Ith_S(y,0)));
+    return 1;
+  } else {
+    printf("  Initial ERKStepEvolve call successful\n");
+  }
+
+
+  /* Reset state to analytical solution at dTout, evolve to 2*dTout and check result */
+  t = T0 + dTout;
+  N_VConst(ytrue(t), y);
+  retval = ERKStepReset(arkode_mem, t, y);
+  check_retval(&retval, "ERKStepReset", 1);
+  retval = ERKStepEvolve(arkode_mem, t + dTout, y, &t, ARK_NORMAL);
+  if (check_retval(&retval, "ERKStepEvolve", 1)) return 1;
+  if (check_ans(y, t, RCONST(0.001), RCONST(0.000001))) {
+    printf("  Second ERKStepEvolve call had insufficient accuracy\n");
+    printf("    t = %" GSYM "\n", t);
+    printf("    y = %" GSYM "\n", NV_Ith_S(y,0));
+    printf("    ytrue = %" GSYM "\n", ytrue(t));
+    printf("    |y-ytrue| = %" GSYM "\n", SUNRabs(ytrue(t)-NV_Ith_S(y,0)));
+    return 1;
+  } else {
+    printf("  Second ERKStepEvolve call successful\n");
+  }
+
+  /* Reset state to analytical solution at 3*dTout, evolve to 4*dTout and check result */
+  t = T0 + RCONST(3.0)*dTout;
+  N_VConst(ytrue(t), y);
+  retval = ERKStepReset(arkode_mem, t, y);
+  check_retval(&retval, "ERKStepReset", 1);
+  retval = ERKStepEvolve(arkode_mem, t + dTout, y, &t, ARK_NORMAL);
+  if (check_retval(&retval, "ERKStepEvolve", 1)) return 1;
+  if (check_ans(y, t, RCONST(0.001), RCONST(0.000001))) {
+    printf("  Third ERKStepEvolve call had insufficient accuracy\n");
+    printf("    t = %" GSYM "\n", t);
+    printf("    y = %" GSYM "\n", NV_Ith_S(y,0));
+    printf("    ytrue = %" GSYM "\n", ytrue(t));
+    printf("    |y-ytrue| = %" GSYM "\n", SUNRabs(ytrue(t)-NV_Ith_S(y,0)));
+    return 1;
+  } else {
+    printf("  Third ERKStepEvolve call successful\n");
+  }
+
```

Comment:
The three tests above cover the cases of resetting to the same time and a later time. It would be good to have a test that resets to an earlier time as well.

Reply:
Good idea.  I added this as a fourth test for each stepper, and all tests still pass.